oak_runtime/src/node/grpc/invocation.rs

```diff
@@ -16,70 +16,54 @@
 
 //! Functionality for handling gRPC method invocations.
 
+pub use crate::proto::oak::invocation::GrpcInvocation as Invocation;
 use crate::{
-    io::{Receiver, ReceiverExt, Sender, SenderExt},
-    NodeMessage, RuntimeProxy,
+    io::{ReceiverExt, SenderExt},
+    RuntimeProxy,
 };
 use log::error;
 use oak_abi::OakStatus;
-use oak_io::{
-    handle::{ReadHandle, WriteHandle},
-    Decodable, OakError,
-};
+use oak_io::OakError;
 use oak_services::proto::oak::encap::{GrpcRequest, GrpcResponse};
 
-/// A gRPC invocation, consisting of exactly two channels: one to read incoming requests from the
-/// client (wrapped in a [`Receiver`]), and one to write outgoing responses to the client (wrapped
-/// in a [`Sender`]).
-pub struct Invocation {
-    receiver: Receiver<GrpcRequest>,
-    sender: Sender<GrpcResponse>,
-}
-
-impl Decodable for Invocation {
-    fn decode(message: &NodeMessage) -> Result<Self, OakError> {
-        if !message.bytes.is_empty() {
-            error!("Non-empty data field");
-            return Err(OakStatus::ErrInternal.into());
-        }
-        if message.handles.len() != 2 {
-            error!(
-                "Incorrect number of handles received: {} (expected: 2)",
-                message.handles.len()
-            );
-            return Err(OakStatus::ErrInternal.into());
-        }
-        Ok(Self {
-            receiver: Receiver::new(ReadHandle {
-                handle: message.handles[0],
-            }),
-            sender: Sender::new(WriteHandle {
-                handle: message.handles[1],
-            }),
-        })
-    }
-}
-
 impl Invocation {
     pub fn close(self, runtime: &RuntimeProxy) {
-        if let Err(err) = self.receiver.close(runtime) {
-            error!("Failed to close receiver channel in invocation: {:?}", err);
-        }
-        if let Err(err) = self.sender.close(runtime) {
-            error!("Failed to close sender channel in invocation: {:?}", err);
-        }
+        match self.receiver {
+            Some(receiver) => {
+                if let Err(err) = receiver.close(runtime) {
+                    error!("Failed to close receiver channel in invocation: {:?}", err);
+                }
+            }
+            None => error!("No receiver on invocation."),
+        };
+        match self.sender {
+            Some(sender) => {
+                if let Err(err) = sender.close(runtime) {
+                    error!("Failed to close sender channel in invocation: {:?}", err);
+                }
+            }
+            None => error!("No sender on invocation."),
+        };
     }
+
     pub fn receive_request(&self, runtime: &RuntimeProxy) -> Result<GrpcRequest, OakError> {
-        self.receiver.receive(runtime)
+        match &self.receiver {
+            Some(receiver) => receiver.receive(runtime),
+            None => Err(OakError::OakStatus(OakStatus::ErrBadHandle)),
```

Comment:
I thought this should be a similar error to above when no receiver or sender is present. Not exactly the same, but they should be related?

Reply:
The above function logs the error as it does not return a result, where this returns a result, so I thought it would be sufficient to propagate the error to be logged at a higher level. Do you suggest logging here and also returning the error or something else?